### PR TITLE
resources op

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref #1117 

Todo:
- [x] return sensible text representation of `Resource`
- [x] tests

### CODE REVIEW

File: js/resources.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import * as flatbuffers from "./flatbuffers";
+import { assert } from "./util";
+import * as dispatch from "./dispatch";
+
+interface Resource {
+  rid: number;
+  repr: string;
+}
+
+export function resources(): Resource[] {
```
Comment (ry) [inline js/resources.ts]: Should we return a `{ [key: number]: string }` instead?

File: js/resources.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as msg from "gen/msg_generated";
+import * as flatbuffers from "./flatbuffers";
+import { assert } from "./util";
+import * as dispatch from "./dispatch";
+
+interface Resource {
+  rid: number;
+  repr: string;
```
Comment (ry) [inline js/resources.ts]: replace `repr: string` with:

`kind: "stdin" | "stdout" | "stderr" | "fsfile" | ... `

File: js/resources.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    const item = res.resources(i)!;
+
+    resources.push({
+      rid: item.rid()!,
+      repr: item.repr()!
+    });
+  }
+
+  return resources.sort((a, b) => a.rid - b.rid);
+}
```
Comment (ry) [inline js/resources.ts]: Maybe combine all three functions into one, since there is only the sync version?

File: js/resources.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  for (let i = 0; i < res.resourcesLength(); i++) {
+    const item = res.resources(i)!;
+
+    resources.push({
+      rid: item.rid()!,
+      repr: item.repr()!
+    });
+  }
+
+  return resources.sort((a, b) => a.rid - b.rid);
```
Comment (ry) [inline js/resources.ts]: Do the sorting on the rust side.

File: js/resources_test.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  assertEqual(res[2].rid, 2);
+  assertEqual(res[2].repr, "stderr");
+});
+
+testPerm({ net: true }, async function netResources() {
+  const addr = "127.0.0.1:4500";
+  const listener = deno.listen("tcp", addr);
+
+  listener.accept().then(async conn => {
```
Comment (ry) [inline js/resources_test.ts]: Use await

File: js/resources_test.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  });
+
+  const conn = await deno.dial("tcp", addr);
+  const buf = new Uint8Array(1024);
+  await conn.read(buf);
+  conn.close();
+});
+
+test(function fileResources() {
+  deno.readFileSync("package.json");
```
Comment (ry) [inline js/resources_test.ts]: readFileSync shouldn't create any resources. You have to use `deno.open('package.json')`

File: js/resources.ts (hunk 1 above)
Comment (bartlomieju) [reply js/resources.ts]: Sure we can do it, then there is no need for `Resource` interface and sorting. I just took a hint from #1117 ;)

Comment (ry) [reply js/resources.ts]: I'm not sure actually. @kitsonk @piscisaureus @kevinkassimo any preferences?

File: js/resources_test.ts (hunk 5 above)
Comment (bartlomieju) [reply js/resources_test.ts]: I took it from `net_test.ts`, IMHO it simpler to use `.then` here, otherwise we need some helper to do the `dial`.

Comment (ry) [reply js/resources_test.ts]: I see. Ok.

File: js/resources.ts (hunk 1 above)
Comment (ry) [reply js/resources.ts]: Discussed with @piscisaureus out of band. Let's go with `{ [key: number]: string }`

**Comment by bartlomieju** [comment]
@ry I updated PR, however another problem surfaced. In `stdioResources` test, actual number of resources is 4, somehow there is an open file. Any idea what might cause that?

**Comment by ry** [comment]
@bartlomieju just compare before and after - perhaps there is a file that doesn't get closed in the tests - the purpose of this is to figure out about those things

**Comment by bartlomieju** [comment]
It seams it fails exactly for that reason on AppVeyor. Looks like `package.json` was already open.

File: js/resources_test.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import { test, testPerm, assert, assertEqual } from "./test_util.ts";
+import * as deno from "deno";
+
+test(function stdioResources() {
```
Comment (ry) [inline js/resources_test.ts]: prefix all of the test names with `resources`.... so call this one `resourcesStdio`

The reason for this is to easily filter tests by module name.

File: js/resources_test.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  const conn = await deno.dial("tcp", addr);
+  const buf = new Uint8Array(1024);
+  await conn.read(buf);
+  conn.close();
+});
+
+test(function fileResources() {
+  const resourcesBefore = deno.resources();
+  deno.open("package.json");
```
Comment (ry) [inline js/resources_test.ts]: Please use `tests/hello.txt`. We've used package.json a lot in the past, but I want to move away from it.

File: js/resources_test.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  const resourcesAfter = deno.resources();
+
+  // check that exactly one new resource (file) was added
+  assertEqual(
+    Object.keys(resourcesAfter).length,
+    Object.keys(resourcesBefore).length + 1
+  );
+  const newRid = Object.keys(resourcesAfter).find(key => {
+    return !resourcesBefore.hasOwnProperty(key);
```
Comment (ry) [inline js/resources_test.ts]: s/key/rid/

File: src/resources.rs [hunk 10]
```diff
@@ -58,6 +58,30 @@ enum Repr {
...
+pub fn get_resource_table_entries() -> Vec<(i32, String)> {
+  let table = RESOURCE_TABLE.lock().unwrap();
+
+  let tuples = table
+    .iter()
+    .map(|(key, value)| (key.clone(), inspect_repr(&value)))
+    .collect();
+
+  tuples
+}
```
Comment (ry) [inline src/resources.rs]: Can you please add a rust test here. Here's a start:
```
#[test]
fn test_get_resource_table_entries() {
  assert_eq!(get_resource_table_entries().len(), 3);
  // Maybe add some other asserts?
}
```

File: src/ops.rs [hunk 11]
```diff
@@ -1288,3 +1290,49 @@ fn op_metrics(
...
+
+  ok_future(serialize_response(
+    cmd_id,
+    builder,
+    msg::BaseArgs {
+      inner: Some(inner.as_union_value()),
+      inner_type: msg::Any::ResourcesRes,
+      ..Default::default()
+    },
+  ))
```
Comment (ry) [inline src/ops.rs]: Nice work!

Comment (bartlomieju) [reply src/ops.rs]: Thanks!

File: src/resources.rs (hunk 10 above)
Comment (bartlomieju) [reply src/resources.rs]: Right now I added TODO, I need some time to figure out how to run Tokio's functions in tests.

Comment (ry) [reply src/resources.rs]: I don't think the above test should work? I don't think you need to interact with tokio?

**Comment by bartlomieju** [comment]
ping @ry

File: js/resources_test.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    assertEqual(Object.values(res).filter(r => r === "tcpStream").length, 2);
+
+    conn.close();
+    listener.close();
+  });
+
+  const conn = await deno.dial("tcp", addr);
+  const buf = new Uint8Array(1024);
+  await conn.read(buf);
+  conn.close();
```
Comment (ry) [inline js/resources_test.ts]: It's possible that the callback doesn't get called at all. Please add a counter which gets incremented in the callback, and check here
`assertEqual(counter, 1)`

Also you can remove ` await conn.read(buf);` - I think that has no effect here.

File: src/resources.rs [hunk 13]
```diff
@@ -58,6 +58,36 @@ enum Repr {
...
+
+#[test]
+fn test_get_resource_table_entries() {
+  assert_eq!(get_resource_table_entries().len(), 3);
+  // TODO: add asserts for add_fs_file, add_tcp_listener, add_tcp_stream
```
Comment (ry) [inline src/resources.rs]: Maybe check some of the entries? 
```
let entries = get_resource_table_entries();
assert_eq!(entries[0], (0, "stdin")
```

Comment (bartlomieju) [reply src/resources.rs]: That's a good idea, but HashMap doesn't guarantee that first entry will be the first one added, does it?

Using this:
```
#[test]
fn test_get_resource_table_entries() {
  let entries = get_resource_table_entries();
  assert_eq!(entries.len(), 3);
  assert_eq!(entries[0], (0, String::from("stdin")));
  assert_eq!(entries[1], (1, String::from("stdout")));
  assert_eq!(entries[2], (2, String::from("stderr")));
}
```
I get this:
```
---- resources::test_get_resource_table_entries stdout ----
thread 'resources::test_get_resource_table_entries' panicked at 'assertion failed: `(left == right)`
  left: `(2, "stderr")`,
 right: `(0, "stdin")`', ../../src/resources.rs:76:3
```

Comment (ry) [reply src/resources.rs]: Call `sort()` inside of `get_resource_table_entries()`

File: js/resources_test.ts (hunk 12 above)
Comment (bartlomieju) [reply js/resources_test.ts]: You were right about that, I reworked it to use `await`, let me know if it looks good.

**Pull request merged**